x/lightclient/keeper/hook_listener.go

```diff
@@ -63,13 +62,15 @@ func (hook rollappHook) AfterUpdateState(
 	// [hStart-1..,hEnd) is correct because we compare against a next validators hash
 	for h := stateInfo.GetStartHeight() - 1; h < stateInfo.GetLatestHeight(); h++ {
 		if err := hook.validateOptimisticUpdate(ctx, rollappId, client, seq, stateInfo, h); err != nil {
-			if errorsmod.IsOf(err, gerrc.ErrFault) {
-				// TODO: should double check this flow when implementing hard fork
-				break
-			}
-			return errorsmod.Wrap(err, "validate optimistic")
+			return errorsmod.Wrap(err, "validate optimistic update")
 		}
 	}
+	// we now verified everything up to and including stateInfo.GetLatestHeight()-1
+	// so we should prune everything up to stateInfo.GetLatestHeight()-1
+	if err := hook.k.PruneSignersBelow(ctx, rollappId, stateInfo.GetLatestHeight()); err != nil {
+		return errorsmod.Wrap(err, "prune signers")
+	}
+
```

Comment:
this won't prune if client is marked canonical, which mean someone can get stuck unbonding

also you still have a signer removal in validateOptimisticUpdate

I recommend my 2 liner pr

Reply:
the issue I had with the doing it on the canonical set, is that it needs `stateInfo.GetStartHeight() - 1` which was pruned in your 2 liner
also the canonical setting already doing the optimistic valdation

Reply:
"this won't prune if client is marked canonical, which mean someone can get stuck unbonding
how this can happen? it will be cleared on the next state update
I think it's fine for our current assumptions. it can be refactored later on